### PR TITLE
python QR code benchmark

### PR DESCRIPTION
**Pull request by AleksandrPanov**

### CODE REVIEW

File: python_benchmarks/QRcodes/QR.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    gl_qr_detect = 0
+    
+    for dir in listDirs:
+        imgsPath = findImagesPath(dir)
+        qr_count = 0
+        qr_detect = 0
+        for imgPath in imgsPath:
+            labelPath = imgPath[:-3] + "txt"
+            gold_corners = getCorners(labelPath)
+            if gold_corners.shape[0] == 1 or True:
```
Comment (asmorkalov) [inline python_benchmarks/QRcodes/QR.py]: Why?

File: python_benchmarks/QRcodes/QR.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import glob
```
Comment (alalek) [inline python_benchmarks/QRcodes/QR.py]: Item 1 of OpenCV coding style guide: https://github.com/opencv/opencv/wiki/Coding_Style_Guide

> All the file names are written in lower case for better compatibility with both POSIX and Windows.

So,

- `qr_codes`
- `qr.py`
- `readme.md` or prefer `README.md` to highlight it (exception is possible as it is not used in other places)

File: python_benchmarks/QRcodes/QR.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import sys
+
+import cv2 as cv
+import numpy as np
+
+accuracy = 10
```
Comment (alalek) [inline python_benchmarks/QRcodes/QR.py]: Parameter or constant?

File: python_benchmarks/qr_codes/qr.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env python
+
+"""qr.py
+Usage example:
+python qr.py -o out.yaml -p qrcodes/detection
+-H, --help - show help
+-o, --output - output file (default out.yaml)
+-p, --path - input dataset path (default qrcodes/detection)
+-a, --accuracy - input accuracy (default 47)
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py]: what is accuracy and what 47 means?

File: python_benchmarks/qr_codes/qr.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+-alg - input alg (default OpenCV)
+-m, --metric - input metric (default ~)
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py]: What is `~`?

File: python_benchmarks/qr_codes/qr.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    return ret, corners
+
+
+def find_images_path(dir_path):
+    images = glob.glob(dir_path+'/*.jpg')
+    images += glob.glob(dir_path+'/*.png')
+    return images
+
+
+def get_corners(label_path):
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py]: I propose to add `ground_truth` or `gold` to the name to stress it.

File: python_benchmarks/qr_codes/qr.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    f = open(label_path, "r")
+    corners = []
+    for line in f.readlines():
+        try:
+            f_list = [float(i) for i in line.split(" ")]
+            corners += f_list
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py:94]: I propose to add at least array size check to validate integrity and parsing issues.

File: python_benchmarks/qr_codes/qr.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        if dist > accuracy:
+            one_corners = np.flip(one_corners, 0)
+            dist = min(dist, get_distance(gold_corner, one_corners))
+            for i in range(0, 3):
+                if dist > accuracy:
+                    one_corners = np.roll(one_corners, 1, 0)
+                    dist = min(dist, get_distance(gold_corner, one_corners))
+                else:
+                    return dist
+    return dist
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py:135]: It should be max distance, but not the last one in the loop.

File: python_benchmarks/qr_codes/qr.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+import cv2 as cv
+import numpy as np
+
+
+def init_qr_detector_open_cv():
+    return cv.QRCodeDetector()
+
+
+def init_wechat_qr_detector_open_cv(path_to_model):
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py]: why open_cv?

File: python_benchmarks/qr_codes/qr.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env python
+
+"""qr.py
+Usage example:
+python qr.py -o out.yaml -p qrcodes/detection
+-H, --help - show help
+-o, --output - output file (default out.yaml)
+-p, --path - input dataset path (default qrcodes/detection)
+-a, --accuracy - input accuracy (default 47)
+-alg - input alg (default OpenCV)
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py]: Not implemented.

File: python_benchmarks/qr_codes/README.md [hunk 11]
```diff
@@ -0,0 +1,2 @@
+Download boofcv dataset (https://boofcv.org/notwiki/regression/fiducial/qrcodes_v3.zip) to this folder.
+Install opencv-python.
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/README.md]: Need to define how to point to WeChat models location.

File: python_benchmarks/qr_codes/qr.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        elif type_detector == self.TypeDetector.opencv_wechat:
+            self.detector = cv.wechat_qrcode_WeChatQRCode(path_to_model + "detect.prototxt",
+                                                          path_to_model + "detect.caffemodel",
+                                                          path_to_model + "sr.prototxt",
+                                                          path_to_model + "sr.caffemodel")
+        else:
+            raise TypeError("this type_detector isn't supported")
+
+    def detect(self, image):
+        if self.type_detector == self.TypeDetector.opencv:
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py]: Why do you need single class with branches. Dedicated classes + duck typing solves it better.

Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: fixed

File: python_benchmarks/qr_codes/README.md (hunk 11 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/README.md]: added info to README

File: python_benchmarks/qr_codes/qr.py (hunk 10 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: fixed

File: python_benchmarks/qr_codes/qr.py (hunk 9 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: fixed

File: python_benchmarks/qr_codes/qr.py (hunk 4 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: fixed

File: python_benchmarks/qr_codes/qr.py (hunk 5 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: fixed

File: python_benchmarks/QRcodes/QR.py (hunk 3 above)
Comment (AleksandrPanov) [reply python_benchmarks/QRcodes/QR.py]: fixed

File: python_benchmarks/qr_codes/README.md [hunk 13]
```diff
@@ -0,0 +1,13 @@
+
+To run the benchmark:
+1. Download boofcv dataset (https://boofcv.org/notwiki/regression/fiducial/qrcodes_v3.zip) to path_folder.
+2. Install opencv-python.
+3. Use "python qr.py -alg opencv -o out.yaml -p path_folder/qrcodes/detection" to run benchmark.
+
+
+To run opencv wechat algorithm:
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/README.md]: OpenCV and WeChat: https://www.wechat.com/

File: python_benchmarks/qr_codes/README.md [hunk 14]
```diff
@@ -0,0 +1,13 @@
+
+To run the benchmark:
+1. Download boofcv dataset (https://boofcv.org/notwiki/regression/fiducial/qrcodes_v3.zip) to path_folder.
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/README.md]: BoofCV: http://boofcv.org/index.php?title=Main_Page

File: python_benchmarks/qr_codes/README.md [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+2. Install opencv-python.
+3. Use "python qr.py -alg opencv -o out.yaml -p path_folder/qrcodes/detection" to run benchmark.
+
+
+To run opencv wechat algorithm:
+1. Follow the steps from the previous part.
+2. Install opencv-contrib-python.
+3. Build OpenCV with OpenCV contrib (https://docs.opencv.org/4.x/db/d05/tutorial_config_reference.html)
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/README.md]: One of 2 or 3.

File: python_benchmarks/qr_codes/README.md [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+4. Find files detect.prototxt, detect.caffemodel, sr.prototxt, sr.caffemodel and set path_to_model.
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/README.md]: I propose to add direct links to the files to readme.

File: python_benchmarks/qr_codes/README.md [hunk 17]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/README.md:3]: I propose to add intro phrase like:
"The benchmark aims to check QR codes detection and decoding with classic OpenCV implementation from objdetect module and dnn based implementation in opencv_contrib repo contributed by the WeChat team."

File: python_benchmarks/qr_codes/qr.py (hunk 4 above)
Comment (asmorkalov) [reply python_benchmarks/qr_codes/qr.py]: It's still there.

File: python_benchmarks/qr_codes/qr.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+from enum import Enum
+
+import numpy as np
+import cv2 as cv
+
+
+class DetectorQR:
+    TypeDetector = Enum('TypeDetector', 'opencv opencv_wechat')
+
+    def __init__(self, type_detector=TypeDetector.opencv, path_to_model="./"):
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py]: It looks like `type_detector` is redundant and used in factory method only. I propose to drop it here and in inherited classes.

File: python_benchmarks/QRcodes/QR.py (hunk 1 above)
Comment (AleksandrPanov) [reply python_benchmarks/QRcodes/QR.py]: fixed

File: python_benchmarks/QRcodes/QR.py (hunk 2 above)
Comment (AleksandrPanov) [reply python_benchmarks/QRcodes/QR.py]: fixed

File: python_benchmarks/qr_codes/qr.py (hunk 4 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: set default `accuracy` as 20 pixels. This value for this dataset (with large images) is a good marker.

File: python_benchmarks/qr_codes/qr.py (hunk 6 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: changed `get_corners` to `get_gold_corners`

File: python_benchmarks/qr_codes/qr.py (hunk 7 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py:94]: `return np.array(corners).reshape(-1, 4, 2)` made array size check.
Do you suggest adding a manual check?

File: python_benchmarks/qr_codes/qr.py (hunk 8 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py:135]: `get_norm_to_rotate_qr` finds the smallest distance between the gold QR and all possible rotations of the found QR code.

File: python_benchmarks/qr_codes/qr.py (hunk 18 above)
Comment (AleksandrPanov) [reply python_benchmarks/qr_codes/qr.py]: type_detector was removed from classes

File: python_benchmarks/qr_codes/qr.py (hunk 7 above)
Comment (asmorkalov) [reply python_benchmarks/qr_codes/qr.py:94]: I propose to add explicit check with verbose error message.

File: python_benchmarks/qr_codes/qr.py [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+    fs = cv.FileStorage(path, cv.FILE_STORAGE_READ)
+    root = fs.root()
+    for image in root.keys():
+        if image.split('_')[0] == "img":
+            image_category = image.split("_")[-2]
+            image_info = root.getNode(image)
+            corners = image_info.getNode("corners").mat()
+            decoded_info = image_info.getNode("decoded_info")
+            if not decoded_info.empty():
+                for i in range(decoded_info.size()):
```
Comment (asmorkalov) [inline python_benchmarks/qr_codes/qr.py:148]: Please remove the dead code.

**Review by asmorkalov — APPROVED** [review]
:+1: LGTM!

File: python_benchmarks/QRcodes/QR.py (hunk 2 above)
Comment (AleksandrPanov) [reply python_benchmarks/QRcodes/QR.py]: Now `qr.py` and `README.md`

**Pull request merged**